### PR TITLE
Namespacing part 2

### PR DESCRIPTION
**Pull request by timwoj**
This should catch the remainder of the core types in the namespacing work. I wanted to get this in before we cut 3.2. There are related PRs on the bifcl and binpac repos that I will link into this one. I attached what the namespacing looks like at this point.

Next up is work on the sub-namespaces (analyzer, file_analysis, etc.)

[parsed.txt](https://github.com/zeek/zeek/files/4827972/parsed.txt)

### CODE REVIEW

**Comment by timwoj** [comment]
Rebased the branch on master.

File: src/Val.cc [hunk 1]
```diff
@@ -379,61 +381,61 @@ void Val::ValDescribeReST(ODesc* d) const
...
-void Val::SetID(zeek::detail::ID* id)
+void Val::SetID(detail::ID* id)
 	{
 	delete [] bound_id;
 	bound_id = id ? copy_string(id->Name()) : nullptr;
 	}
 #endif
 
-bool Val::WouldOverflow(const zeek::Type* from_type, const zeek::Type* to_type, const Val* val)
+bool Val::WouldOverflow(const ::zeek::Type* from_type, const ::zeek::Type* to_type, const Val* val)
```
Comment (jsiwek) [inline src/Val.cc:396]:
```suggestion
bool Val::WouldOverflow(const zeek::Type* from_type, const zeek::Type* to_type, const Val* val)
```

File: src/Conn.h [hunk 2]
```diff
@@ -29,8 +29,14 @@ class RuleHdrTest;
 class Specific_RE_Matcher;
 class RuleEndpointState;
 class EncapsulationStack;
-class Val;
-class RecordVal;
+
+ZEEK_FORWARD_DECLARE_NAMESPACED(Val, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(RecorVal, zeek);
```
[the remark this answers is not in the document]
Comment (timwoj) [reply src/Conn.h:34]: I guess that line isn't needed at all since `RecordVal` is needed a couple of lines later in that file.

**Comment by timwoj** [comment]
> Func and FuncPtr may be better in zeek than detail ? Would guess we want to try to keep the function call/invoke API stable for people.

But leave BuiltinFunc and ScriptFunc in detail?

> BroObj -> zeek::Obj

This brings up a larger question if we should just take this opportunity to rename all of the `Bro` names.

File: src/List.h [hunk 3]
```diff
@@ -26,8 +26,9 @@
 #include <cassert>
 #include "util.h"
 
-// TODO: this can be removed in v3.1 when List::sort() is removed
-typedef int (*list_cmp_func)(const void* v1, const void* v2);
+//enum class [[deprecated("Remove in v4.1. Use zeek::ListOrder instead.")]] ListOrder : int { ORDERED, UNORDERED };
```
[the remark this answers is not in the document]
Comment (timwoj) [reply src/List.h:29]: Yep, forgot to remove it when I added the deprecated version.

**Comment by jsiwek** [comment]
> But leave BuiltinFunc and ScriptFunc in detail?

Think so.  Most people probably don't care to know/use those, they'll just want to lookup a base `Func` by name and `Invoke` it.

> This brings up a larger question if we should just take this opportunity to rename all of the `Bro` names.

Good catch, guess some of the others had already slipped by, but yeah, throwing out ideas for renaming things I see on the list:

* `Bro::UID` -> `zeek::UID`
* `BroFile` -> `zeek::File` (maybe detail, not sure)
* `BroSubstring` -> `zeek::detail::Substring`
* `BroString` -> `zeek::String`
* `Brofiler` -> `zeek::detail::ScriptProfiler`
* `binpac::Bro`* -> didn't look at where those actually live, but just Zeek* versions may make sense
* `bro_broker::`* -> probably most go in `zeek::broker::detail::`* except `zeek::broker::Manager` / `zeek::broker::mgr` (from `broker_mgr`)
* `BroExprTag` -> `zeek::detail::ExprTag`

Definitely see some other `bro_` stuff not in the list, like `bro_random()`, `bro_path()`, `bro_int_t` that warrant rename eventually.  Not that scoping all ese into this PR matters, but for ones it already touched, should just do the Bro -> Zeek conversion now.

**Comment by timwoj** [comment]
Fixed all of the comments and did a bit of additional renaming:
zeek::detail::Func/FuncPtr -> zeek::Func/FuncPtr
zeek::detail::BroFunc -> zeek::detail::ScriptFunc (and fixed the enum)
BroObj -> Obj
BroString -> String

I left BroString.h in place because moving to String.h causes a ton of errors from compilers that are case insensitive when it comes to looking up include files when it conflicts with string.h.

**Pull request merged**